nixos/modules/services/development/zammad.nix

```diff
@@ -0,0 +1,217 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.services.zammad;
+  serviceConfig = {
+    Type = "simple";
+    Restart = "always";
+
+    User = "zammad";
+    Group = "zammad";
+    PrivateTmp = true;
+    StateDirectory = builtins.baseNameOf cfg.dataDir;
+    WorkingDirectory = cfg.dataDir;
+
+    EnvironmentFile = cfg.secretsFile;
+  };
+  env = {
+    RAILS_ENV = "production";
+    NODE_ENV = "production";
+    RAILS_SERVE_STATIC_FILES="true";
+    RAILS_LOG_TO_STDOUT="true";
+  };
+in {
+
+  options = {
+    services.zammad = {
+      enable = lib.mkEnableOption "Zammad, a web-based, open source user support/ticketing solution.";
+
+      package = lib.mkOption {
+        type = lib.types.package;
+        default = pkgs.zammad;
+        defaultText = "pkgs.zammad";
+        description = "Zammad package to use.";
+      };
+
+      dataDir = lib.mkOption {
+        type = lib.types.path;
+        default = "/var/lib/zammad";
+        description = ''
+          Path to a folder that will contain Zammad working directory.
+        '';
+      };
+
+      host = lib.mkOption {
+        type = lib.types.str;
+        default = "127.0.0.1";
+        example = "192.168.23.42";
+        description = "Host address.";
+      };
+
+      port = lib.mkOption {
+        type = lib.types.int;
+        default = 3000;
+        description = "Web service port.";
+      };
+
+      websocketPort = lib.mkOption {
+        type = lib.types.int;
+        default = 6042;
+        description = "Websocket service port.";
+      };
+
+      dbName = lib.mkOption {
+        type = lib.types.str;
+        default = "zammad";
+        description = "The name of the database to use.";
+      };
+
+      dbUsername = lib.mkOption {
+        type = lib.types.str;
+        default = "zammad";
+        description = "The username to use to connect to the database.";
+      };
+
+      secretsFile = lib.mkOption {
+        type = lib.types.nullOr lib.types.path;
+        default = null;
+        description = ''
+          Path of a file containing secrets the format of EnvironmentFile as
+          described by systemd.exec(5). You must to define:
+            - PGPASSWORD
+            - SECRET_KEY_BASE
+          SECRET_KEY_BASE can be generated using:
+            ruby -e "require 'securerandom'; puts SecureRandom.hex(64)"
+        '';
+      };
+    };
+
+  };
+
+  config = lib.mkIf cfg.enable {
+
+    networking.firewall.allowedTCPPorts = [
+      config.services.zammad.port
+      config.services.zammad.websocketPort
+    ];
+
+    users.users.zammad = {
+      isSystemUser = true;
+      home = cfg.dataDir;
+      group = "zammad";
+    };
+
+    users.groups.zammad = {};
+
+    services.postgresql = {
+      enable = true;
```

Comment:
that doesn't look like a good thing to be honest. it can be an *option* (`createLocalDatabase` or similar), but enabling a local postgres by default is a big hammer when remote servers could also be used. if you use the unix socket for postgres you could avoid the whole username/password dance and use unix peer authentication instead. :)

Reply:
Okay. I'm trying to rewrite the database bit using https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/mail/sympa.nix as a model

Reply:
That looks like a great source to use :+1: